### ISSUE TITLE
Remove links the same way you create them

### ISSUE THREAD

**Issue by Atakku** [Ax9D/pw-viz#12]
It seems to be much more intuitive to just remove links by dragging them over the same path again, as you would create them.

**Comment by Atakku** [comment]
I have alt used for dragging around windows in my WM, and helvum removes links the way I described above

**Comment by jansol** [comment]
With how it currently works, that would make it impossible to link the same port to multiple nodes.

However making this work more like Blender would solve your problem (and help make the Linux ecosystem marginally more consistent):

* Drag starting from an output port always creates a new link
* Drag starting from an input port visually detaches that end of the link and
  * if drag ends on the same port where it started, do nothing
  * if drag ends on a different input port, create a new link between that port and the dragged link's starting point, then delete the original link
  * if drag ends on "empty space", delete the link

**Comment by Atakku** [comment]
Sorry, blender wasn't the best analogy. How about helvum, it has the behavior I described

**Comment by jansol** [comment]
Ah okay, seems I'd misunderstood your first post. So performing the same drag gesture between two ports that are already linked would delete the link. That does indeed work even for an arbitrary number of links since any given pair of ports can only have one link between them.

The main shortcoming of the way it is done in Helvum is that there is practically no visual hint about what you are currently doing until you are hovering the mouse cursor on the destination node (assuming the drag gesture went in the right direction, if not, nothing happens), and deleting links seems to show the same cursor as adding them. With proper (live) UI feedback it could be made decently usable I guess. And adding and deleting nodes should not care about the direction of the drag gesture, i.e. draggin from an output to an input and from an input to an output should both yield the same result.

Now that I think of it, the Blender way does not work as is anyway, since in Blender it does not make sense to have multiple links connected to the same input port, whereas in Pipewire that is very common. I do however find that Blender feels closer to the analogy of plugging in and unplugging physical cables. Not sure what the best way to keep that would be. Having a key that toggles between creating and deleting nodes is nice (especially if the key was configurable) but it does not solve the ambiguity when multiple links connect to the same port. Being able to select a link by clicking and presssing delete would be easy and unambiguous. Your solution would also work but requires a fair amount of UI feedback work.

**Comment by Ax9D** [comment]
I agree with @jansol on this, dragging back a link completely along the same path would be very tedious for complex graphs,  dragging back a link from an input port to delete it makes more sense imo
I have a patch for egui_nodes which enables which enables this behaviour:
![drag_back](https://user-images.githubusercontent.com/40864118/149015251-f77c4453-cd41-4d9c-b8c7-658abcb8ae6b.gif)
and this works without holding Alt ofcourse

**Comment by jansol** [comment]
How does it choose which link to disconnect if there are multiple links to the same input?

**Comment by mriise** [comment]
@jansol I came across this issue yesterday, when multiple links are put into the same input it will select whichever seems to be on "top" i think. So trying to remove a single link you may have to remove multiple lines before you are able to grab the line you wanted, which is annoying. Being able to drag from somewhere along the middle of the line to disconnect would allow for moving nodes around to clear the area around a line so there is easier access to grab just the one you intend.

**Comment by actionless** [comment]
> How does it choose which link to disconnect if there are multiple links to the same input?

mb by right click on the connection node to show dropdown menu which will allow to choose which of the nodes is 'active'
